Content.Client/GameObjects/Components/Kitchen/ReagentGrinderBoundUserInterface.cs

```diff
@@ -0,0 +1,316 @@
+            _beakerVisualContents?.Clear();
+            _menu?.Dispose();
+        }
+
+        protected override void UpdateState(BoundUserInterfaceState state)
+        {
+            base.UpdateState(state);
+            if (!(state is ReagentGrinderInterfaceState cState))
+            {
+                return;
+            }
+
+            _menu.BeakerContentBox.EjectButton.Disabled = !cState.HasBeakerIn;
+            _menu.ChamberConentBox.EjectButton.Disabled = cState.ChamberContents.Length <= 0;
+            RefreshContentsDisplay(cState.ReagentQuantities, cState.ChamberContents, cState.HasBeakerIn);
+
+        }
+
+
+        protected override void ReceiveMessage(BoundUserInterfaceMessage message)
+        {
+            base.ReceiveMessage(message);
+            switch(message)
+            {
+                case SharedReagentGrinderComponent.ReagentGrinderWorkStartedMessage workStarted:
+                    _menu.GrindButton.Disabled = true;
+                    _menu.GrindButton.Modulate = workStarted.IsJuiceIntent ? Color.White : Color.Green;
+                    _menu.JuiceButton.Disabled = true;
+                    _menu.JuiceButton.Modulate = !workStarted.IsJuiceIntent ? Color.White : Color.Green;
+                    break;
+                case SharedReagentGrinderComponent.ReagentGrinderWorkCompleteMessage doneMessage:
+                    _menu.GrindButton.Disabled = false;
+                    _menu.JuiceButton.Disabled = false;
+                    _menu.GrindButton.Modulate = Color.White;
+                    _menu.JuiceButton.Modulate = Color.White;
+                    break;
+            }
+        }
+
+        private void RefreshContentsDisplay(Solution.ReagentQuantity[] reagents, EntityUid[] containedSolids, bool isBeakerAttached)
+        {
+
+            //Much of this component's interface will just be ripped straight from microwave...
+            _chamberVisualContents.Clear();
+            _menu.ChamberConentBox.BoxContents.Clear();
+            for (var j = 0; j < containedSolids.Length; j++)
+            {
+                if (!_entityManager.TryGetEntity(containedSolids[j], out var entity))
+                {
+                    return;
+                }
+                if (entity.Deleted)
+                {
+                    continue;
+                }
+
+                Texture texture;
+                if (entity.TryGetComponent(out IconComponent iconComponent))
+                {
+                    texture = iconComponent.Icon?.Default;
+                }
+                else if (entity.TryGetComponent(out SpriteComponent spriteComponent))
+                {
+                    texture = spriteComponent.Icon?.Default;
+                }
+                else { continue; }
+
+                var solidItem = _menu.ChamberConentBox.BoxContents.AddItem(entity.Name, texture);
+                var solidIndex = _menu.ChamberConentBox.BoxContents.IndexOf(solidItem);
```

Comment:
I feel like you should be able to get around using IndexOf using either a dictionary or an array